File: LinkedList/SinglyLinkedList/Medium/FirstLoopNode.java
package Medium;

import java.util.HashSet;
import x.*;

public class FirstLoopNode {
    public static void main(String[] args) {
        LinkedList l1 = new LinkedList();
        l1.append(1);
        l1.append(2);
        l1.append(3);
        l1.append(4);
        l1.append(5);

        l1.head.next.next.next.next.next = l1.head.next;
        getFirstLoopNode(l1);
    }

    static void getFirstLoopNode(LinkedList list) {
        Node cur = list.head;
        // 1(cur)->2->3->4->5->2; hashset = 1, 2, 3, 4, 5
        HashSet<Node> list_set = new HashSet<Node>();

        while (cur != null) {
            if (list_set.contains(cur)) {
                // loop exits
                System.out.println(cur.val);
                break;
            } else {
                // still no loop
                list_set.add(cur);
                cur = cur.next;
            }
        }
    }
}

File: LinkedList/SinglyLinkedList/Medium/FirstLoopNode.js
// starting node of the loop
import { LinkedList } from "../x/LinkedList.js";

let list = new LinkedList([1, 2, 3, 4, 5, 6]);
list.head.next.next.next = list.head;

const firstLoopNode = (list) => {
    let cur = list.head;
    let prev = null;
    let set = new Set();

    while (cur != null) {
        // check if the cur.val is in set
        if (set.has(cur)) {
            // loop exists
            return prev;
        } else {
            set.add(cur);
        }
        prev = cur;
        cur = cur.next;
    }

    return "No loop detected";
};

console.log(firstLoopNode(list));
